### PR TITLE
DEPR: Timestamp.freq

### LINKED ISSUE #15146: DEPR: remove freq attribute of Timestamp?
xref https://github.com/pandas-dev/pandas/issues/15141#issuecomment-273184725

In theory these should be completely fungible. If so, we should remove the ``freq`` attribute
of ``Timestamps`` and instead return ``Periods`` in those instances.

```
In [2]: t = Timestamp('2010-01-03 00:00:00', freq='W-SUN')

In [3]: t
Out[3]: Timestamp('2010-01-03 00:00:00', freq='W-SUN')

In [4]: t.to_period()
Out[4]: Period('2009-12-28/2010-01-03', 'W-SUN')

In [5]: t.to_period().to_timestamp(how='start')
Out[5]: Timestamp('2009-12-28 00:00:00')

In [6]: t.to_period().to_timestamp(how='end')
Out[6]: Timestamp('2010-01-03 00:00:00')
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #15146
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The motivation here is that I've concluded that DatetimeArray.freq _must_ be removed, so for internal consistency Timestamp.freq should be removed too.

The main user-facing behavior this will affect is Timestamp.is_month_start, is_month_end, is_quarter_start, is_quarter_end, is_year_start, is_year_end, for which I guess we'll offer users an alternative(?)

### CODE REVIEW

**Comment by mroeschke** [comment]
> The main user-facing behavior this will affect is Timestamp.is_month_start, is_month_end, is_quarter_start, is_quarter_end, is_year_start, is_year_end, for which I guess we'll offer users an alternative(?)

I think all of these can be defined without frequency. Looks like there's already "no frequency" code path for these

```
        if self.freq is None:
            # fast-path for non-business frequencies
            return self.day == 1 and self.month % 3 == 1
```

File: pandas/conftest.py [hunk 1]
```diff
@@ -78,6 +82,13 @@ def pytest_runtest_setup(item):
         pytest.skip("skipping high memory test since --run-high-memory was not set")
 
 
+def pytest_collection_modifyitems(config, items):
```
Comment (jreback) [inline pandas/conftest.py:85]: does this show up in a lot of places?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you link to the deprecation removal issue as well

**Comment by jbrockmendel** [comment]
> I think all of these can be defined without frequency. Looks like there's already "no frequency" code path for these

You're right.  The potential issue is if users want to get the old (i.e. current) behavior that as if there were a freq.  I guess we can tell them to do `DatetimeIndex([ts], freq=freq).whatever`

File: pandas/conftest.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/conftest.py:85]: Yes.  I was putting `pytestmark = ...` in dozens of files and barely making a dent before putting it here instead

**Comment by jorisvandenbossche** [comment]
The current PR seems to generate a lot of warnings (from looking at the test changes) that shouldn't be propagated to the user? 
(eg boxing a DatetimeIndex, converting to object dtyp, ... And I also noticed that accessing a single element from a DatetimeIndex raises the warning, and the DatetimeIndex repr, ...) 

> The main user-facing behavior this will affect is Timestamp.is_month_start, is_month_end, is_quarter_start, is_quarter_end, is_year_start, is_year_end, for which I guess we'll offer users an alternative(?)

Can you show with an example what is actually changing here?

**Comment by jbrockmendel** [comment]
closing to clear the queue

**Pull request closed without merging**